### ISSUE TITLE
Decide on and document recommended style for docstrings in coding standard

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2029]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2029|
|Type|enhancement|
|Created|2006-08-29 07:11:35Z|
|Branch|https://github.com/twisted/twisted/tree/docstrings-style-2029|

Many docstrings are written in a person voice, e.g. "I am Jack's twisted colon".  This is fun, but it's also a little silly.

Either we should:
 1. Write _all_ of our docstrings in that style OR
 1. Write _none_ of our docstrings in that style.

I vote for 2.

In either case, we should update the documentation standard, and fix the existing non-conforming docstrings.

<details><summary>Searchable metadata</summary>

```
trac-id__2029 2029
type__enhancement enhancement
reporter__jml jml
priority__high high
milestone__ 
branch__branches_docstrings_style_2029 branches/docstrings-style-2029
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1156835495000000 1156835495000000
changetime__1224707667000000 1224707667000000
version__None None
owner__ 
cc__thijs cc__exarkun cc__dash cc__radix cc__glyph
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Originally, all of Twisted's docstrings were in this style.  It's a Smalltalk convention; have a look at Squeak's API documentation.  For the one-sentence summary this can be awkward (as recently demonstrated), but it can make longer docstrings easier to write, because it provides a consistent convention for referring to the thing being documented, instead of having to refer to it sometimes by name, sometimes as "this function", "this method", or straining to avoid any self-reference by using the passive voice.

However, apparently this leads to a similar problem with "I perform", "I have", "I am" and "I contain".  You can still get into problem without the first-person voice though: "Performs function X".  "Has feature X".  "Contains feature X".  I think the first priority here is to come up with some convention for the module-summary style to describe its relation to its functionality.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

My previous comment is sort of meandering.  The main thing I meant to say was:

First-person vs. non-first-person is one of the least interesting aspects of this discussion.  I think we need to document conventions for how to describe:

 * a package's relationship to its functionality,
 * a module's relationship to its contents,
 * a class's relationship to its methods and attributes, and
 * a function or method's relationship to its behavior.

I think parameters are pretty well covered by @param and friends, so I haven't listed those.  Similarly, the conventions I'm talking about aren't about listing classes, methods and functions, they're about how to generally describe the collections.

One thing we can start the discussion with: the words "implements" and "provides" have to be omitted, because those mean very specific things in Zope Interface terms, and we should only use those specific technical meanings within the documentation.  Maybe that's a separate ticket.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra removed owner|
|-|-|

I vote for 2 as well. If more people add their opinion here, we can (finally) put it in a document.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

Hum, I thought this was decided a long time ago.  At least, I've been deleting first-person style docs near code I touch. ;)

Discussion over, I think.  Option 2 wins, let's document it if it needs to be documented somewhere and close this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Agreed.  I still like first-person but it has seemed pretty clearly unpopular for a long time :).

In regards to my earlier comment, I still don't think it's just a simple choice between "first person" or "not first person", but how to use a consistent style to describe the class's relationship to its functionality.  The style I have adopted in place of this:
```
#!py
# option A
class Ninja:
    """
    I am a warrior specializing in a variety of unorthodox arts of war.
    """
    def attack(self, someone):
        """
        Attack C{someone} with my shiruken.
        """
```
is this:
```
#!py
# option B
class Ninja:
    """
    A L{Ninja} is a warrior specializing in various unorthodox arts of war.
    """
    def attack(self, someone):
        """
        Attack C{someone} with this L{Ninja}'s shiruken.
        """
```
A brief survey of some other code suggests that another frequently-used style is this:
```
#!py
# option C
class Ninja:
    """
    a warrior specializing in various unorthodox arts of war.
    """
    def attack(self, someone):
        """
        Attack C{someone} with the ninja's shiruken.
        """
```

I strongly prefer option B.  Option A is unpopular, and although option C is used in various places (I looked at `twisted.internet.interfaces` and `twisted.mail.imap4`), it requires the use of sentence fragments — many docstrings don't have a subject — and is often ambiguous.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

I like complete sentences.  _A L{Ninja} ..._ is fine with me, although I think the link in a class's docstring to the class itself isn't adding a lot of value. :)  I also don't mind _This class ..._, though.  I don't think we need to enforce one or the other of these choices, but I'm happy to say that docstrings should use complete sentences, proper punctuation, etc.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

The link isn't useful as such, certainly.  It's just a habit so that it's properly formatted as a class in the documentation and the reader gets all the visual hints that indicate "I am referring to this precise class, not a general idea".

I prefer _A L{Thing}_ to _This class_ because I generally want to describe the behavior of instances of a class, and _Each instance of this class_ sounds awkward, and "this class" is often subtly inaccurate.  This is especially true when you need to start describing interfaces that the class itself provides, behavior of classmethods, etc.  I do sometimes use _this module_ or _this function_ though, since no such distinction or ambiguity exists in that case.

I'm opining here not to suggest rooting out all incomplete sentences and replacing them or making all documentation slavishly follow this guideline, but because I think the documentation of this decision should include an example that we can follow, and I'd like something like option B for that example.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Replying to [glyph](#note_8):
> The link isn't useful as such, certainly.  It's just a habit so that it's properly formatted as a class in the documentation and the reader gets all the visual hints that indicate "I am referring to this precise class, not a general idea".

Indeed, and actually your points about "visual hints" and "precise class" are a good ones.  I (we) shouldn't think of `L{}` as meaning "present a link to a page documenting the contained API in the html output of processing this documentation" but rather "this is an FQPN".  The resulting markup is the job of the markup generation tool, and if I think links to class X shouldn't be included in class X's docstring, then I can file a ticket against pydoctor to have it generate a different (but probably still visually distinctive) kind of markup for that case.

>
> I prefer _A L{Thing}_ to _This class_ because I generally want to describe the behavior of instances of a class, and _Each instance of this class_ sounds awkward, and "this class" is often subtly inaccurate.  This is especially true when you need to start describing interfaces that the class itself provides, behavior of classmethods, etc.  I do sometimes use _this module_ or _this function_ though, since no such distinction or ambiguity exists in that case.
>
> I'm opining here not to suggest rooting out all incomplete sentences and replacing them or making all documentation slavishly follow this guideline, but because I think the documentation of this decision should include an example that we can follow, and I'd like something like option B for that example.

Sounds good to me.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

(In [24927]) Branching to 'docstrings-style-2029'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set owner to @thijstriemstra|
|-|-|

  1. The `foo2bar` docstring is using `C{}` where it should be using `L{}`.  It would be a much better example if it talked about converting `L{Foo}`s to `L{Bar}`s.
  2. The `foo2bar` docstring also refers to `L{FooException()}` where it should be referring to `L{FooException}`.
  3. Similarly, I'd prefer the `Ninja` docstring to be the one in my "option B" comment, above.

With these modifications it would be fine to merge.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [25027]) Merge docstrings-style-2029: Documented recommended style for docstrings in the coding standard.

Author: thijs
Reviewer: glyph, exarkun
Fixes: [#2029](https://github.com/twisted/twisted/issues/2029)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
